client-tools/Shell/RequestFuzzer.cpp

```diff
@@ -0,0 +1,278 @@
+    switch (charOp) {
+      case CharOperation::ADD_STRING:
+        generateRandReadableAsciiString(input);
+        break;
+      case CharOperation::ADD_INT_32: {
+        input.append(std::to_string(generateRandInt32()));
+        break;
+      }
+      default:
+        TRI_ASSERT(false);
+    }
+  }
+}
+
+void RequestFuzzer::randomizeLineOperation(uint32_t numIts) {
+  uint32_t its = numIts;
+  while (its--) {
+    LineOperation lineOp = static_cast<LineOperation>(
+        _randContext.mt() % LineOperation::MAX_LINE_OP_VALUE);
+    switch (lineOp) {
+      case LineOperation::INJECT_RAND_BYTE_IN_LINE: {
+        if (_stringLines.size() > 1) {
+          uint32_t initLinePos;
+          if (generateRandNumWithinRange<uint32_t>(0, 99) > 0) {
+            initLinePos = 1;
+          } else {
+            initLinePos = 0;
+          }
+          uint32_t randPos;
+          randPos = generateRandNumWithinRange<uint32_t>(
+              initLinePos, _stringLines.size() - 1);
+          if (_stringLines[randPos].find("Content-length") ==
+                  std::string::npos &&
+              _reqHasBody) {
+            _stringLines.emplace_back("Content-length:" +
+                                      std::to_string(_reqBodySize));
+          }
+          generateRandAsciiChar(_stringLines[randPos]);
+        }
+        break;
+      }
+      case LineOperation::COPY_LINE: {
+        if (_stringLines.size() > 1) {
+          _stringLines.emplace_back(
+              _stringLines.at(generateRandNumWithinRange<uint32_t>(
+                  1, _stringLines.size() - 1)));
+        }
+        break;
+      }
+      case LineOperation::ADD_LINE: {
+        uint32_t numHeaderFields =
+            generateRandNumWithinRange<uint32_t>(0, _wordListForKeys.size());
+        for (uint32_t i = 0; i < numHeaderFields; ++i) {
+          uint32_t keyPos = generateRandNumWithinRange<uint32_t>(
+              0, _wordListForKeys.size() - 1);
+          std::string keyName;
+          std::string value;
+          if (!(_wordListForKeys[keyPos] == "random") &&
+              !(_wordListForKeys[keyPos] == "Content-length")) {
+            keyName = _wordListForKeys[keyPos];
+          } else {
+            randomizeCharOperation(keyName, 1);
+          }
+          randomizeCharOperation(value, 1);
+          _keysAndValues[keyName] = value;
+        }
+        if (_reqHasBody) {
+          _keysAndValues.try_emplace("Content-length",
+                                     std::to_string(_reqBodySize));
+        }
+        for (auto const& [key, value] : _keysAndValues) {
+          _stringLines.emplace_back(std::move(key) + ":" + std::move(value));
+        }
+        break;
+      }
+      default:
+        TRI_ASSERT(false);
+    }
+  }
+}
+
+void RequestFuzzer::randomizeHeader(std::string& header,
+                                    std::optional<size_t> const payloadSize) {
+  if (payloadSize.has_value()) {
+    _reqBodySize = payloadSize.value();
```

Comment:
Ok, randomizeHeader is capturing the body size... Why? Shouldn't it just deal with the header?

Reply:
It needs to know the byte size of the body for the content-length, but this can be somehow refactored